### PR TITLE
Glass Floors Can Break When Jumped On

### PR DESCRIPTION
**Pull request by Thunder12345**
## About The Pull Request

Glass floors have a chance of breaking when someone *jump emotes on top of one.

## Why It's Good For The Game

It's a funny interaction for enhancing slapstick comedy.

## Changelog
:cl:
add: NanoTrasen no longer guarantees that glass floors are safe to jump on. Please exercise caution.
/:cl:

### CODE REVIEW

**Comment by Sealed101** [comment]
finally

**Review by VladinXXV — CHANGES_REQUESTED** [review]
All my other problems with this aside, your fragility numbers came straight out of thin air and are ridiculous. See below for reference.

MATERIAL - WINDOW INTEGRITY - FLOOR FRAGILITY
Glass - 50 - 80
Reinforced Glass - 75 -5
Plasmaglass - 200 - 50
Reinforced Plasmaglass - 500 - 3

In no world is plasmaglass easier to break than reinforced glass, nor is it ever anywhere NEAR reinforced plasmaglass. Frankly, the suggestion that anyone is breaking plasmaglass - let alone REINFORCED plasmaglass - by JUMPING on it is completely and utterly absurd.

**Comment by Jacquerel** [comment]
This is funny, but emotes are spammable to the extent that the chance doesn't really matter and this imposes constraints on where mappers can use this which did not previoulsy exist.
Possibly I could be convinced otherwise but I do not think that _any_ glass floor should do this aside from the basic "glass plating", any other variant should be impervious to jumping.

Also this should be a component you can apply to any flooring, and if you actually want to commit to this as a mechanic it should trigger when objects which hurt people when they land on them land on that tile as well (pushing a piano from the top z level of icebox onto the middle level of icebox should drop it through the glass floor onto the bottom level of icebox).
And perhaps even when shot with bullets, so you can intentionally have someone fall down a level.

**Comment by Vect0r2** [comment]
I like it!

**Comment by necromanceranne** [comment]
Maybe if you didn't jump on the glass floor, you wouldn't have come crashing down onto my banquet table, you dumb cat.

File: code/modules/mob/living/emote.dm [hunk 1]
```diff
@@ -209,9 +209,22 @@
...
+	message = "jumps up and down."
 	hands_use_check = TRUE
 
+/datum/emote/living/jump/run_emote(mob/living/user, params, type_override, intentional)
+	. = ..()
+	var/turf/open/current = get_turf(user)
+	if(!isopenturf(current))
+		return
+	if(prob(current.jump_fragility))
+		playsound(current, SFX_SHATTER, 70, TRUE)
```
Comment (Mothblocks) [inline code/modules/mob/living/emote.dm:221]:
```suggestion
		playsound(current, SFX_SHATTER, 70, vary = TRUE)
```

File: code/modules/mob/living/emote.dm [hunk 2]
```diff
@@ -209,9 +209,22 @@
...
+		user.visible_message(
+				span_warning("[user] crashes through [current]!"),
+				span_userdanger("You crash through [current]!"),
+			)
```
Comment (Mothblocks) [inline code/modules/mob/living/emote.dm:225]:
```suggestion
		user.visible_message(
			span_warning("[user] crashes through [current]!"),
			span_userdanger("You crash through [current]!"),
		)
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/emote.dm [hunk 3]
```diff
@@ -209,9 +209,22 @@
 /datum/emote/living/jump
 	key = "jump"
 	key_third_person = "jumps"
-	message = "jumps!"
+	message = "jumps up and down."
 	hands_use_check = TRUE
 
+/datum/emote/living/jump/run_emote(mob/living/user, params, type_override, intentional)
+	. = ..()
+	var/turf/open/current = get_turf(user)
```
Comment (Jacquerel) [inline code/modules/mob/living/emote.dm:217]: Does this work from inside a locker? _Should_ it work from inside a locker?
One of those cases where `user.loc` might actually be more appropriate.

**Comment by Ryusenshu** [comment]
Can glass floors even be removed normally? we dont even have a tool tip for it as far as i know

**Comment by Jacquerel** [comment]
> Can glass floors even be removed normally? we dont even have a tool tip for it as far as i know

I think they can't be deconstructed with tools (an RCD still works) but that smells like a bug rather than an intended feature. I don't think we're supposed to have unremovable floors.

File: code/modules/mob/living/emote.dm (hunk 3 above)
Comment (Thunder12345) [reply code/modules/mob/living/emote.dm:217]: I didn't test that, but no reason it wouldn't work from inside a locker. As for whether it should, does it enhance or detract from the comedy if someone bounces a locker up and down to drop it through the floor?

**Comment by Thunder12345** [comment]
> Maybe if you didn't jump on the glass floor, you wouldn't have come crashing down onto my banquet table, you dumb cat.

That was not OSHA compliant

**Comment by Farquaar** [comment]
Reinforced glass shouldn't break when jumped on. That's kinda dumb.

**Comment by tralezab** [comment]
normal glass floors break when jumping enough
reinforced glass floors bnreak when falling from further zlevels
can't break plasmaglass

that would be reasonable

**Comment by Jacquerel** [comment]
> normal glass floors break when jumping enough reinforced glass floors bnreak when falling from further zlevels can't break plasmaglass
> 
> that would be reasonable

This would also be my preference alongside implementing this as a component.
If you are worried about componentising it making it annoying for subtypes, you would keep your "break chance" var and only apply the component if it has a `> 0` value.

File: code/modules/mob/living/emote.dm (hunk 3 above)
Comment (optimumtact) [inline code/modules/mob/living/emote.dm:217]: in my opinion this should only work if their direct loc is a floor.

File: code/modules/mob/living/emote.dm [hunk 4]
```diff
@@ -209,9 +209,22 @@
...
+	if(!isopenturf(current))
+		return
+	if(prob(current.jump_fragility))
+		playsound(current, SFX_SHATTER, 70, vary = TRUE)
+		user.visible_message(
+			span_warning("[user] crashes through [current]!"),
+			span_userdanger("You crash through [current]!"),
+		)
+		current.ScrapeAway(2, flags = CHANGETURF_INHERIT_AIR)
```
Comment (optimumtact) [inline code/modules/mob/living/emote.dm:226]: does this behave if there's no actual z level below the glass floor?

**Review by optimumtact — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/emote.dm (hunk 4 above)
Comment (Mothblocks) [reply code/modules/mob/living/emote.dm:226]: This is what we do for explosions as well, so it should

**Comment by san7890** [comment]
> > Can glass floors even be removed normally? we dont even have a tool tip for it as far as i know
> 
> I think they can't be deconstructed with tools (an RCD still works) but that smells like a bug rather than an intended feature. I don't think we're supposed to have unremovable floors.

this is intentional as of #66301 and any attempt to revert this feature must directly address this PR. glass floors should be analogous to plating for multi-z balance reasons (not indestructible, just at least should follow the same rules as plating) imo

**Comment by Jacquerel** [comment]
San’s comment leads me to one additional concern. 
It is not yet in wide use but may become more so as more maps incorporate multi-z; so does this change allow you to place one single glass floor tile over the top of Reinforced Plating and then jump on it to create a hole, bypassing the deconstruction process completely?

**Comment by Thunder12345** [comment]
Welp, apparently the only place a lot of people's minds go as soon as they see a comedy feature is powergaming, they're why we can't have nice things sometimes

**Pull request closed without merging**